nscplugin/src/main/scala/scala/scalanative/nscplugin/NirGenStat.scala

```diff
@@ -218,6 +229,354 @@ trait NirGenStat { self: NirGenPhase =>
+      }
+
+      def genLazyClassInstantiationMethod(exprBuf: ExprBuffer,
+                                          ctors: Seq[global.Symbol]): Val = {
+        val applyMethodSig =
+          Sig.Method("apply", Seq(jlObjectRef, jlObjectRef))
+
+        // Constructors info is an array of Tuple2 (tpes, inst), where:
+        // - tpes is an array with the runtime classes of the constructor arguments.
+        // - inst is a function, which accepts an array with tpes and returns a new
+        //   instance of the class.
+        val ctorsInfo = exprBuf.arrayalloc(Type.Array(tuple2Ref),
+                                           Val.Int(ctors.length),
+                                           unwind(curFresh))
+
+        // For each (public) constructor C, generate a lambda responsible for
+        // initialising and returning an instance of the class, using C.
+        for ((ctor, ctorIdx) <- ctors.zipWithIndex) {
+          val ctorSig     = genMethodSig(ctor)
+          val ctorArgsSig = ctorSig.args.map(_.mangle).mkString
+
+          ReflectiveInstantiationInfo += new ReflectiveInstantiationBuffer(
+            fqSymId + ctorArgsSig)
+          val reflInstBuffer = ReflectiveInstantiationInfo.last
+
+          // Lambda generation consists of generating a class which extends
+          // scala.runtime.AbstractFunction1, with an apply method that accepts
+          // the list of arguments, instantiates an instance of the class by
+          // forwarding the arguments to C, and returns the instance.
+          withFreshExprBuffer { exprBuf =>
+            val body = {
+              // first argument is this
+              val thisArg = Val.Local(curFresh(), Type.Ref(reflInstBuffer.name))
+              // second argument is parameters sequence
+              val argsArg = Val.Local(curFresh(), Type.Array(jlObjectRef))
+              exprBuf.label(curFresh(), Seq(thisArg, argsArg))
+
+              // Extract and cast arguments to proper types.
+              val argsVals =
+                (for ((arg, argIdx) <- ctorSig.args.tail.zipWithIndex) yield {
+                  val elem =
+                    exprBuf.arrayload(jlObjectRef,
+                                      argsArg,
+                                      Val.Int(argIdx),
+                                      unwind(curFresh))
+                  // If the expected argument type can be boxed (i.e. is a primitive
+                  // type), then we need to unbox it before passing it to C.
+                  Type.box.get(arg) match {
+                    case Some(_) =>
+                      // first, cast the primitive type to java.lang.Number
+                      val num = exprBuf.as(jlNumberRef, elem, unwind(curFresh))
+                      // then, convert to the desired type
+                      val conv = exprBuf.method(num,
+                                                Type.primConvSig(arg),
+                                                unwind(curFresh))
```

Comment:
I think that we also need to take care of shrinking conversions. If a constructor expects a `Short` and someone provides e.g. `42`, which is an `Int`, we still have a mismatch.

I think that we need something more generic, like:

```scala
  private def primConversionIfRequired(arg: Any,
                                       paramType: Class[_]): Any = {
    paramType match {
      case java.lang.Byte.TYPE =>
        arg.asInstanceOf[Number].byteValue()
      case java.lang.Short.TYPE =>
        arg.asInstanceOf[Number].shortValue()
      case java.lang.Integer.TYPE =>
        arg.asInstanceOf[Number].intValue()
      case java.lang.Long.TYPE =>
        arg.asInstanceOf[Number].longValue()
      case java.lang.Float.TYPE =>
        arg.asInstanceOf[Number].floatValue()
      case java.lang.Double.TYPE =>
        arg.asInstanceOf[Number].doubleValue()
      case _ => arg
    }
  }
```
It is basically what I was doing in the codegen, moved to the library side.

Reply:
> This won't work for booleans and characters.

Why do we need to convert `Boolean`s and `Character`s? They are not numeric types.

Reply:
If you follow the link to the spec I gave, you'll see that narrowing conversions shouldn't be accepted, and that characters can be converted to `int` and bigger.

As validation, here is a transcript of a REPL on the JVM to show that this is indeed what happens:
```scala
scala> case class Foo(x: Int, y: java.lang.Long)
defined class Foo

scala> val c = classOf[Foo].getDeclaredConstructors()(0)
c: java.lang.reflect.Constructor[_] = public Foo(int,java.lang.Long)

scala> c.newInstance(1, 2L)
res4: Object = Foo(1,2)

scala> c.newInstance(1, 2)
java.lang.IllegalArgumentException: argument type mismatch
  at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
  at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
  at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
  at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
  ... 34 elided

scala> c.newInstance(1, 2.toShort)
java.lang.IllegalArgumentException: argument type mismatch
  at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
  at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
  at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
  at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
  ... 34 elided

scala> c.newInstance(1L, 2L)
java.lang.IllegalArgumentException: argument type mismatch
  at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
  at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
  at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
  at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
  ... 34 elided

scala> c.newInstance(1.toShort, 2L)
res8: Object = Foo(1,2)

scala> c.newInstance(1.4, 2L)
java.lang.IllegalArgumentException: argument type mismatch
  at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
  at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
  at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
  at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
  ... 34 elided

scala> c.newInstance('A', 2L)
res10: Object = Foo(65,2)

scala> case class Foo(x: Int, y: java.lang.Long, z: Short, c: Char)
defined class Foo

scala> val c = classOf[Foo].getDeclaredConstructors()(0)
c: java.lang.reflect.Constructor[_] = public Foo(int,java.lang.Long,short,char)

scala> c.newInstance('A', 2L, 'A', 'B')
java.lang.IllegalArgumentException: argument type mismatch
  at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
  at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
  at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
  at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
  ... 34 elided

scala> c.newInstance('A', 2L, 5, 'B')
java.lang.IllegalArgumentException: argument type mismatch
  at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
  at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
  at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
  at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
  ... 34 elided

scala> c.newInstance('A', 2L, 5.toByte, 'B')
res13: Object = Foo(65,2,5,B)

scala> c.newInstance('A', 2L, 5.toShort, 'B')
res14: Object = Foo(65,2,5,B)

scala> c.newInstance('A', 2L, 5.toShort, 5.toShort)
java.lang.IllegalArgumentException: argument type mismatch
  at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
  at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
  at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
  at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
  ... 34 elided

scala> c.newInstance('A', 2L, 5.toShort, 5.toByte)
java.lang.IllegalArgumentException: argument type mismatch
  at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
  at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
  at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
  at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
  ... 34 elided
```

Reply:
> 
> 
> > This won't work for booleans and characters.
> 
> Why do we need to convert `Boolean`s and `Character`s? They are not numeric types.

`Boolean`s need not be converted, but the code that is currently in the PR would try to cast `Boolean`s to `Number`s anyway, even if the expected target type was `Boolean`. Or maybe I misunderstood the code. For `Char` we in fact need to convert them to numeric types, as I showed above.

Reply:
Okay, thanks for the clarification.

I got confused from a test case from Scala.js. In the `testClassCtorWithArgs` test, the `Short` constructor is requested and I though it was "converted" from the `Int` one. However I now see the presence of the `VC` class. The current implementation does not support such a scenario.

Reply:
The current implementation in Scala.js seems incomplete about the conversions. We'll need to fix that when I get the chance.

Reply:
So, if I understand this correctly, instantiating a `Short` constructor with `42`, which is [supported](https://github.com/scala-js/scala-js/blob/3534fda873127e5da1badbefa376ce0c7d951d03/test-suite/js/src/test/scala/org/scalajs/testsuite/library/ReflectTest.scala#L179) in Scala.js, should not be supported in Native.

Reply:
Therefore, if I understand this correctly, calling a `Short` constructor with a value of `42` (which is an `Int`), which is [supported by Scala.js](https://github.com/scala-js/scala-js/blob/3534fda873127e5da1badbefa376ce0c7d951d03/test-suite/js/src/test/scala/org/scalajs/testsuite/library/ReflectTest.scala#L179), should not be supported by Native?

I suppose that Scala.js treats numeric values differently, in a JS-specific way. I am a little confused as to which semantics we would like to "follow". In C for example, it is OK (to some extent) to do all kinds of conversions. Why not do the same in Native and why is the behaviour of Scala.js different in that aspect?

Reply:
> Therefore, if I understand this correctly, calling a Short constructor with a value of 42 (which is an Int), which is supported by Scala.js, should not be supported by Native?

That's right.

> I suppose that Scala.js treats numeric values differently, in a JS-specific way. I am a little confused as to which semantics we would like to "follow". In C for example, it is OK (to some extent) to do all kinds of conversions. Why not do the same in Native and why is the behaviour of Scala.js different in that aspect?

We have to follow the JVM semantics first. Copying C or Scala.js is secondary.

Scala.js treats boxed primitive values differently than the JVM across the board (see https://www.scala-js.org/doc/semantics.html) for JS-specific reasons, and is therefore not a gold standard to follow if the JVM does not agree. Basically in Scala.js we *cannot* tell apart a `j.l.Integer` with value 42 from a `j.l.Integer` with value 42, so we're forced to accept it.

> On the other hand, Scala seems to accept the above scenario:

That's with *compile-time* adaptation. Java does the same in that situation. But here we are concerned with *run-time* adaptation as performed by Java reflection, since that is the capability we are trying to emulate.